### PR TITLE
NCCL2 support

### PR DESCRIPTION
**Pull request by anaruse**
This PR is for supporting NCCL2.

With this PR, your model can be trained with multiple GPUs both in fp32 and fp16 using NCCL2. Note that you need to install https://github.com/cupy/cupy/pull/363 to use this PR.

### CODE REVIEW

**Comment by okuta** [comment]
jenkins, test this please.

**Comment by okuta** [comment]
@anaruse Please resolve conflicts.

File: chainer/training/updaters/multiprocess_parallel_updater.py [hunk 1]
```diff
@@ -357,9 +443,19 @@ def scatter_params(link, array):
...
-    offset = 0
-    for _, param in sorted(link.namedparams()):
-        next_offset = offset + param.size
-        param.data = array[offset:next_offset].reshape(param.data.shape)
-        offset = next_offset
-    assert array.size == offset
+    return _scatter(link, array, "data")
+
+
+def get_nccl_data_type(dtype):
```
Comment (okuta) [inline chainer/training/updaters/multiprocess_parallel_updater.py]: How about `_getnccl_data_type` ?

File: chainer/training/updaters/multiprocess_parallel_updater.py [hunk 2]
```diff
@@ -335,19 +354,86 @@ def gather_params(link):
...
+        dtypes[i] = 0  # fp32
+        if param.dtype == numpy.float16:
+            dtypes[i] = 1  # fp16
+        info[i + 1] = info[i] + param.size
+        i += 1
+    if i != num:
+        raise()
+    info[0] = num
+
+    ptrs = cuda.to_gpu(ptrs, stream=cuda.Stream.null)
```
Comment (okuta) [inline chainer/training/updaters/multiprocess_parallel_updater.py]: `stream` option was deprecated.

File: chainer/training/updaters/multiprocess_parallel_updater.py [hunk 3]
```diff
@@ -299,14 +314,18 @@ def _gather(link, target):
...
             ptrs[i] = d.data.ptr
+        dtypes[i] = 0  # fp32
+        if param.dtype == numpy.float16:
+            dtypes[i] = 1  # fp16
         info[i + 1] = info[i] + param.size
         i += 1
     info[0] = num
 
     ptrs = cuda.to_gpu(ptrs, stream=cuda.Stream.null)
+    dtypes = cuda.to_gpu(dtypes, stream=cuda.Stream.null)
```
Comment (okuta) [inline chainer/training/updaters/multiprocess_parallel_updater.py]: `stream` option is deprecated.

**Comment by anaruse** [comment]
@okuta Thank you for your comments. The branch was fixed based on your comments.

**Comment by okuta** [comment]
jenkins, test this please.

**Comment by okuta** [comment]
LGTM!

**Pull request merged**